src/Datadog.Trace/RuntimeMetrics/RuntimeMetricsWriter.cs

```diff
@@ -0,0 +1,131 @@
+using System;
+using System.Diagnostics;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using Datadog.Trace.Vendors.StatsdClient;
+
+namespace Datadog.Trace.RuntimeMetrics
+{
+    internal class RuntimeMetricsWriter : IDisposable
+    {
+        private static readonly string[] GcCountMetricNames = { "runtime.dotnet.gc.count.gen0", "runtime.dotnet.gc.count.gen1", "runtime.dotnet.gc.count.gen2" };
+        private static readonly string[] CompactingGcTags = { "compacting_gc:true" };
+        private static readonly string[] NotCompactingGcTags = { "compacting_gc:false" };
+
+        private readonly int _delay;
+
+        private readonly IDogStatsd _statsd;
+        private readonly Timer _timer;
+        private readonly Process _currentProcess;
+
+#if NETCOREAPP
+        private readonly RuntimeEventListener _listener;
+#endif
+
+        private readonly Timing _contentionTime = new Timing();
+
+        private TimeSpan _previousUserCpu;
+        private TimeSpan _previousSystemCpu;
+
+        private long _contentionCount;
+
+        public RuntimeMetricsWriter(IDogStatsd statsd, int delay)
+        {
+            _delay = delay;
+            _statsd = statsd;
+            _timer = new Timer(_ => PushEvents(), null, delay, delay);
+            _currentProcess = Process.GetCurrentProcess();
+            AppDomain.CurrentDomain.FirstChanceException += FirstChanceException;
```

Comment:
This may also need to be inside a try/catch block since the event is `SecurityCritical` and we wouldn't have permissions in the rare scenario of running .NET Framework with limited permissions

Reply:
Done, nice catch!